### PR TITLE
#823 Implemented assignPhotonAngles function

### PR DESCRIPTION
**Pull request by sethdigel**
The assignPhotonAngles function generates directions (in the form of dx/dz and dy/dz slopes) for an existing photon array.  The generated directions are uniformly distributed in solid angle for inclination angles between the obscuration angle and the FOV size (defined in terms of the f-ratio).  An associated testing script was also implemented.

### CODE REVIEW

**Comment by rmjarvis** [comment]
Note: All the C++ layer stuff is also in #827, which includes real unit tests for it all, so probably better to review that part on #831 rather than here.  But Seth needed this functionality so I implemented it here first.  The files photon_array.py and test_photon_array.py are really what this PR is about.

**Comment by rmjarvis** [comment]
@sethdigel I reworked the front end of this code slightly in anticipation of the design in #827.  Now this is a class FRatioAngles that can `applyTo` a photon array, rather than be a method of PhotonArray directly.  That way it will be easier to have several different options for how we assign directions to the photons.

File: galsim/photon_array.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                            will be automatically constructed for you. [default: None]
+    """
+    def __init__(self, fratio, obscuration=0.0, rng=None):
+
+        if fratio < 0:
+            raise ValueError("The f-ratio must be positive.")
+        if obscuration < 0 or obscuration >= 1:
+            raise ValueError("The obscuration fraction must be between 0 and 1.")
+        if rng is None:
+            ud = galsim.UniformDeviate()
```
Comment (rmjarvis) [inline galsim/photon_array.py]: This line is currently uncovered by the tests.  We should have a test that makes FRatioAngles without a specified seed.  Probably also one that uses an existing BaseDeviate rather than an integer seed.  We don't necessarily need to do the full test of the distribution with these, especially the default one, but tests of min/max values are probably worthwhile.

File: galsim/photon_array.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                            linear dimension. [0., 1.).  [default: 0.0]
+    @param rng              A random number generator to use or None, in which case an rng
+                            will be automatically constructed for you. [default: None]
+    """
+    def __init__(self, fratio, obscuration=0.0, rng=None):
+
+        if fratio < 0:
+            raise ValueError("The f-ratio must be positive.")
+        if obscuration < 0 or obscuration >= 1:
+            raise ValueError("The obscuration fraction must be between 0 and 1.")
```
Comment (rmjarvis) [inline galsim/photon_array.py]: Probably worth adding some assert_raises tests to check these lines work properly.  (They aren't required for our coverage report, but they're worth covering anyway.).  Something like the following:
```
try:
    np.testing.assert_raises(ValueError, FRatioAngles, fratio=-0.3)
    np.testing.assert_raises(ValueError, FRatioAngles, fratio=1.2, obscuration=-0.3)
    np.testing.assert_raises(ValueError, FRatioAngles, fratio=1.2, obscuration=1.0)
    np.testing.assert_raises(ValueError, FRatioAngles, fratio=1.2, obscuration=1.9)
except ImportError:
    pass
```
The try/except wrapper is because these tests technically require nose, which we don't have as an official dependency, so we don't require the user to have it installed.

File: galsim/photon_array.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        fov_angle = np.arctan(0.5 / self.fratio)  # radians
+        obscuration_angle = self.obscuration * fov_angle
+
+        # Generate azimuthal angles for the photons
+        # Set up a loop to fill the array of azimuth angles for now
+        # (The array is initialized below but there's no particular need to do this.)
+        phi = np.zeros(n_photons)
+
+        for i in np.arange(n_photons):
+            phi[i] = self.ud() * 2 * np.pi
```
Comment (rmjarvis) [inline galsim/photon_array.py]: This would be more efficient as
```
phi = self.ud.generate(n_photons) * 2 * np.pi
```
The `generate` function is new in Alex's branch, so once that's merged into master, it would be worth merging it into this branch and using it here.

File: galsim/photon_array.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+        # Generate inclination angles for the photons, which are uniform in sin(theta) between
+        # the sine of the obscuration angle and the sine of the FOV radius
+        sintheta = np.zeros(n_photons)
+
+        for i in np.arange(n_photons):
+            sintheta[i] = np.sin(obscuration_angle) + (np.sin(fov_angle) - \
+                          np.sin(obscuration_angle))*self.ud()
```
Comment (rmjarvis) [inline galsim/photon_array.py]: Likewise, this would be more efficient as
```
sinobs = np.sin(obscuration_angle)
sinfov = np.sin(fov_angle)
sintheta = sinobs + (sinfov - sinobs) * self.ud.generate(n_photons)
```

File: galsim/photon_array.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+        # Assign the directions to the arrays. In this class the convention for the
+        # zero of phi does not matter but it might if the obscuration dependent on
```
Comment (rmjarvis) [inline galsim/photon_array.py]: typo: either "depends" or "is dependent"

File: tests/test_photon_array.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    path, filename = os.path.split(__file__)
+    sys.path.append(os.path.abspath(os.path.join(path, "..")))
+    import galsim
+
+@timer
+def test_photon_angles():
+    """Test the photon_array function
+    """
+    # Make a photon array
+    ud = galsim.UniformDeviate()
```
Comment (rmjarvis) [inline tests/test_photon_array.py]: This should be given a definite seed to make sure tests don't fail randomly at rare intervals.

File: tests/test_photon_array.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    assigner.applyTo(photon_array)
+
+    dxdz = photon_array.getDXDZArray()
+    dydz = photon_array.getDYDZArray()
+
+    # Compare these slopes with the expected distributions (uniform in azimuth 
+    # over all azimiths and uniform in sin(inclination) over the range of
+    # allowed inclinations
+    phi = np.arctan2(dydz,dxdz)
+    sintheta = np.sqrt(1. - np.square(dxdz) - np.square(dydz))
```
Comment (rmjarvis) [inline tests/test_photon_array.py]: I don't think this is quite right.  I think tan(theta) = sqrt(dxdz^2 + dydz^2), which leads to sin(theta) = tan(theta) / sqrt(1+tan(theta)^2).  
However, in the unit test, it's probably worth using the more obvious trig formula:
```
tantheta = np.sqrt(np.square(dxdz) + np.square(dydz))
sintheta = np.sin(np.arctan(tantheta))
```
I'm not sure how much this affects the below tests of the distribution.

File: tests/test_photon_array.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    np.testing.assert_almost_equal(chisqr_phi, 1.05562, 5, \
+        "Distribution in azimuth is not uniform")
+    np.testing.assert_almost_equal(chisqr_sintheta, 0.95920, 5, \
+        "Distribution in sin(inclination) is not uniform")
+
+    # Also check that the values are within the ranges expected
+    np.testing.assert_almost_equal(np.min(phi) + np.pi, 0.000258, 5, \
+        "Azimuth angle range extends to too-small angles")
+    np.testing.assert_almost_equal(np.max(phi) - np.pi, -6.474875e-05, 5, \
+        "Inclination angle range extends to too-large angles")
```
Comment (rmjarvis) [inline tests/test_photon_array.py]: I think these could just check that -pi < phi < pi.  (Also the message on the second one is wrong.)
```
np.testing.assert_less(phi, np.pi, "Azimuth angle range extends to too-large angles")
np.testing.assert_less(-np.pi, phi, "Azimuth angle range extends to too-small angles")
```
Likewise for the below tests of sintheta.

File: tests/test_photon_array.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+                      sintheta_ref) / sintheta_histo.size
+
+    
+    # In the assert_almost_equal tests below, the expected values are defined
+    # for the particular set of directions generated for the seed value specified
+    # above
+    np.testing.assert_almost_equal(chisqr_phi, 1.05562, 5, \
+        "Distribution in azimuth is not uniform")
+    np.testing.assert_almost_equal(chisqr_sintheta, 0.95920, 5, \
+        "Distribution in sin(inclination) is not uniform")
```
Comment (rmjarvis) [inline tests/test_photon_array.py]: I think probably what we really care about here is that it is near 1, so maybe test that 0.9 < chisqr < 1.1 in each case.  Rather than test the exact values.

File: tests/test_photon_array.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    # allowed inclinations
+    phi = np.arctan2(dydz,dxdz)
+    sintheta = np.sqrt(1. - np.square(dxdz) - np.square(dydz))
+
+    phi_histo, phi_bins = np.histogram(phi, bins=100)
+    sintheta_histo, sintheta_bins = np.histogram(sintheta, bins=100)
+
+    phi_ref = phi_histo*0 + float(np.sum(phi_histo))/phi_histo.size
+    sintheta_ref = sintheta_histo*0 + float(np.sum(sintheta_histo)
+                   )/sintheta_histo.size
```
Comment (rmjarvis) [inline tests/test_photon_array.py]: I think you don't need the `*0` bits.  The below calculations should work if the references are floats rather than arrays.

File: galsim/photon_array.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+
+    def applyTo(self, photon_array):
+        """Assign directions to the photons in photon_array."""
+
+        dxdz = photon_array.getDXDZArray()
+        dydz = photon_array.getDYDZArray()
+        n_photons = len(dxdz)
+
+        fov_angle = np.arctan(0.5 / self.fratio)  # radians
```
Comment (rmjarvis) [inline galsim/photon_array.py]: Probably comment with a brief summary of the math for why this is the right formula would be appropriate above this line.

File: tests/test_photon_array.py (hunk 7 above)
Comment (sethdigel) [inline tests/test_photon_array.py]: Thanks, you're right.  I've fixed this on the generation side as well.  I was treating dxdz and dydz like direction cosines

File: tests/test_photon_array.py (hunk 8 above)
Comment (sethdigel) [inline tests/test_photon_array.py]: Thanks.  I've implemented the checks using np.testing.assert_array_less.

File: galsim/photon_array.py (hunk 3 above)
Comment (sethdigel) [inline galsim/photon_array.py]: I've implemented this using self.ud.generate as you suggested.  The syntax is not quite as above because the argument is not n_photons but an array of that size.

**Comment by sethdigel** [comment]
Thanks a lot, Mike.  I've pushed updated versions with changes to address each of your comments.

**Comment by rmjarvis** [comment]
Great.  Looks good Seth.  I added one more thing to the test suite (testing rng=None, ud, or seed), which all passed and got the coverage up to 100%.

Anyone else want to review?

File: galsim/photon_array.py [hunk 12]
```diff
@@ -179,3 +180,68 @@ def applyTo(self, photon_array):
         """Assign wavelengths to the photons sampled from the SED * Bandpass."""
         photon_array.wavelength = self.sed.sampleWavelength(
                 photon_array.size(), self.bandpass, rng=self.rng, npoints=self.npoints)
+
+
+class FRatioAngles(object):
+    """A surface-layer operator that assigns photon directions based on the f/ratio and
+    obscuration.
+
+    Assigns arrival directions at the focal plane for photons, drawing from a uniform
```
Comment (jmeyers314) [inline galsim/photon_array.py]: I think "angle of the FOV" isn't quite right.  Maybe "angle of the pupil"?

File: galsim/photon_array.py [hunk 13]
```diff
@@ -179,3 +180,68 @@ def applyTo(self, photon_array):
...
+
+    def applyTo(self, photon_array):
+        """Assign directions to the photons in photon_array."""
+
+        dxdz = photon_array.getDXDZArray()
+        dydz = photon_array.getDYDZArray()
+        n_photons = len(dxdz)
+
+        # The f/ratio is the ratio of the focal length to the diameter of the aperture of
+        # the telescope.  The angular radius of the field of view is defined by the
```
Comment (jmeyers314) [inline galsim/photon_array.py]: I think FOV is the wrong term here too.  For LSST, the fratio is 1.234, which leads to a maximum inclination angle of ~22 degrees.  `fov_angle` makes me think of the ~1.75 degree radius of the FOV in sky coordinates though.  So I think something like `pupil_angle` makes more sense.

File: galsim/photon_array.py [hunk 14]
```diff
@@ -179,3 +180,68 @@ def applyTo(self, photon_array):
...
+
+        dxdz = photon_array.getDXDZArray()
+        dydz = photon_array.getDYDZArray()
+        n_photons = len(dxdz)
+
+        # The f/ratio is the ratio of the focal length to the diameter of the aperture of
+        # the telescope.  The angular radius of the field of view is defined by the 
+        # ratio of the radius of the aperture to the focal length
+        fov_angle = np.arctan(0.5 / self.fratio)  # radians
+        obscuration_angle = self.obscuration * fov_angle
```
Comment (jmeyers314) [inline galsim/photon_array.py]: I think this is only true under the small angle approximation.  I'm not sure 22 degrees is small enough though for LSST.  Should probably use `np.arctan(0.5 * self.obscuration / self.fratio)` instead.

File: galsim/photon_array.py (hunk 13 above)
Comment (sethdigel) [inline galsim/photon_array.py]: Yes, that is what I should have used.

File: galsim/photon_array.py (hunk 14 above)
Comment (sethdigel) [inline galsim/photon_array.py]: Thanks for this.  I had the mistaken impression that the obscuration fraction was defined as a fraction of the pupil angular radius but I see that in Mike's description it is a fraction of the aperture size.

**Comment by sethdigel** [comment]
I've committed a new version of photon_array.py addressing Josh's comments (and merging in the other updates since yesterday).

**Comment by jmeyers314** [comment]
Thanks Seth.  Looks good to me.
